Modify DirectoryLister interface to pass partition information

abhiseksaikia opened the pull request:
In order to support mutation of partitions, we need partition information during listing time. We are changing the Directory lister interface for the same. We also need to inject additional split related information which has complicated structure and cannot be represented by the existing map structure in HiveSplit. We are going to use existing "extraFileInfo" field of type:byte[] and will leverage properties in customSplitInfo map for information necessary to deserialize the byte[]. 

Test plan - unit test

```
== NO RELEASE NOTE ==
```

NikhilCollooru reviewed (commented):
First commit looks fine. But we should change the commit title and description. This commit is modifying the list api interface right ?

NikhilCollooru commented on presto-hive-common/src/main/java/com/facebook/presto/hive/HiveFileInfo.java:
```diff
@@ -30,26 +32,37 @@
+                ImmutableMap.of());
+    }
+
+    public static HiveFileInfo createHiveFileInfo(LocatedFileStatus locatedFileStatus, Optional<byte[]> extraFileContext, Map<String, String> customSplitInfo)
```
Why do we need customSplitInfo in HiveFileInfo ? i thought we found a way to re-use the existing `byte[] extraFileInfo` for our usecase ?

abhiseksaikia replied:
We will still need to inject custom property (mutation enabled = true) in the split for worker to deserialize the byte[] accordingly.

abhiseksaikia commented:
> First commit looks fine. But we should change the commit title and description. This commit is modifying the list api interface right ?

Thanks for reviewing it! I have squashed the commits and updated the description.

arunthirupathi reviewed (approved):
the change looks good, let me know when you are ready and i can help merge it in!.

arunthirupathi commented:
I did not check the commit message (my mistake). For the next PRs please follow the commit message guidelines here: https://cbea.ms/git-commit/
